nwb_conversion_tools/datainterfaces/axonadatainterface.py

```diff
@@ -0,0 +1,132 @@
+            
+    return header
+
+
+def read_iso_datetime(set_file):
+    """ 
+    Creates datetime object (y, m, d, h, m, s) from .set file header 
+    """
+    with open(set_file, 'r', encoding='cp1252') as f:
+        for line in f:
+            if line.startswith('trial_date'):
+                date_string = re.findall(r'\d+\s\w+\s\d{4}$', line)[0]
+            if line.startswith('trial_time'):
+                time_string = line[len('trial_time')+1::].replace('\n', '')
+
+    return datetime.datetime.strptime(date_string + ', ' + time_string, \
+        "%d %b %Y, %H:%M:%S").isoformat()
+
+
+class AxonaRecordingExtractorInterface(BaseRecordingExtractorInterface):
+    """Primary data interface class for converting a AxonaRecordingExtractor."""
+
+    RX = se.AxonaRecordingExtractor
+
+    @classmethod
+    def get_source_schema(cls):
+        source_schema = {
+            'required': ['filename'],
+            'properties': {
+                'filename': {
+                    'type': 'string',
+                    'format': 'file',
+                    'description': 'Path to Axona files.'
+                }
+            },
+            'type': 'object',
+            'additionalProperties': True
+        }
+        return source_schema
+       
+    def __init__(self, **source_data):
+        super().__init__(**source_data)
+        
+    def get_metadata(self):
+        """Auto-fill as much of the metadata as possible. Must comply with metadata schema."""        
+                
+        # Extract information for specific parameters from .set file
+        # TODO not all are currently used or part of the metadata_schema
+        params_of_interest = [
+            'experimenter', 
+            'comments',
+            'duration', 
+            'sw_version',
+            'tracker_version',
+            'stim_version',
+            'audio_version'
+        ]
+        set_file = self.source_data['filename']+'.set'
+        par = parse_generic_header(set_file, params_of_interest)
+        
+        # Add available metadata
+        # NOTE that this interface is meant to be used within an NWBconverter, 
+        # which contains a much larger metadata_schema. As such a datainterface
+        # metadata by itself does not necessarily validate with its own schema!
+        metadata = super().get_metadata()
+        metadata['NWBFile'] = dict(
+            session_start_time=read_iso_datetime(set_file),
+            session_description=par['comments'],
+            identifier=''.join(random.choices(string.ascii_uppercase + string.digits, k=12)),
+            experimenter=[par['experimenter']]
+        )
+        
+        metadata['Ecephys'] = dict(
+            Device=dict(
+                    name="Axona",
+                    description="Axona DacqUSB, sw_version={}".format(par['sw_version']),
+                    manufacturer="Axona"
+                ),
+            ElectrodeGroup=dict(
+                name='Group0',
```

Comment:
Are there multiple tetrodes? If so, each should be its own group

Reply:
I thought so, too. What threw me off was that making ElectrodeGroup a list of dicts failed the validation, and I wasn't exactly sure how to incorporate it then. Presumably by changing the metadata_schema, but that seems somewhat crude. I am also not sure why lists of dicts are not accepted, since those are used in other examples. For instance [here](https://github.com/catalystneuro/nwb-conversion-tools/blob/master/nwb_conversion_tools/datainterfaces/blackrockdatainterface.py).

And since the datatype is "object" in the schema, list should qualify. This is the relevant part of the metadata_schema:

```
      "ElectrodeGroup": {
        "required": [
          "location",
          "description",
          "device",
          "name"
        ],
        "properties": {
          "name": {
            "description": "the name of this electrode",
            "type": "string"
          },
          "description": {
            "description": "description of this electrode group",
            "type": "string"
          },
          "location": {
            "description": "description of location of this electrode group",
            "type": "string"
          },
          "device": {
            "description": "the device that was used to record from this electrode group",
            "type": "string",
            "target": "pynwb.device.Device"
          }
        },
        "type": "object",
        "additionalProperties": false,
        "tag": "pynwb.ecephys.ElectrodeGroup"
      },
```

This is the error when using a list:

```
---------------------------------------------------------------------------
ValidationError                           Traceback (most recent call last)
<ipython-input-45-27393e1000b0> in <module>
      1 # Validate metadata with metadata_schema from nwbconverter
      2 
----> 3 validate(
      4     instance=converter.get_metadata(),
      5     schema=converter.get_metadata_schema()

/mnt/d/spikeinterface/hussaini-to-nwb/venv/lib/python3.8/site-packages/jsonschema/validators.py in validate(instance, schema, cls, *args, **kwargs)
    932     error = exceptions.best_match(validator.iter_errors(instance))
    933     if error is not None:
--> 934         raise error
    935 
    936 

ValidationError: [{'name': 'Group0', 'location': '', 'device': '', 'description': 'Group 0 electrodes.'}] is not of type 'object'

Failed validating 'type' in schema['properties']['Ecephys']['properties']['ElectrodeGroup']:
    {'additionalProperties': False,
     'properties': {'description': {'description': 'description of this '
                                                   'electrode group',
                                    'type': 'string'},
                    'device': {'description': 'the device that was used to '
                                              'record from this electrode '
                                              'group',
                               'target': 'pynwb.device.Device',
                               'type': 'string'},
                    'location': {'description': 'description of location '
                                                'of this electrode group',
                                 'type': 'string'},
                    'name': {'description': 'the name of this electrode',
                             'type': 'string'}},
     'required': ['location', 'description', 'device', 'name'],
     'tag': 'pynwb.ecephys.ElectrodeGroup',
     'type': 'object'}

On instance['Ecephys']['ElectrodeGroup']:
    [{'description': 'Group 0 electrodes.',
      'device': '',
      'location': '',
      'name': 'Group0'}]
```

Reply:
I see

1. An `object` is mapped to a dictionary in json-schema (see [json-schema docs](https://json-schema.org/understanding-json-schema/reference/object.html))
2. You are right, this should really accept a list of groups. @CodyCBakerPhD and  @luiztauffer , do you know what's going on here? Have we passed multiple groups through in other conversion pipelines? How did you accommodate that?

Reply:
@sbuergers See [Neuroscope](https://github.com/catalystneuro/nwb-conversion-tools/blob/master/nwb_conversion_tools/datainterfaces/neuroscopedatainterface.py#L81-L87) for an example of properly formatted (and schema passing) electrode groups.

Now, there's a lot of parts to that error so I'll start with some of the easier ones. Is that first output the result of asking for the metadata schema through our interface? If so, we'll have to think about a certain field that is 

Please try specifying only 'name' and 'description' in that dictionary in the list and see if it complains then. The schema errors aren't always that specific about _what part_ failed validation and I suspect it's the names of the keys.

@bendichter Yeah, for #2 there the ElectrodeGroups always takes a list of dicts. This is actually [hard enforced on the spikeextractors side](https://github.com/SpikeInterface/spikeextractors/blob/master/spikeextractors/extractors/nwbextractors/nwbextractors.py#L428-L429), but schema validation is failing before it gets there.

Reply:
The schema is formed [here](https://github.com/catalystneuro/nwb-conversion-tools/blob/54608fd49a7eff977cfb9d69d5cd73e90e78e927/nwb_conversion_tools/baserecordingextractorinterface.py#L42), which right now returns the definition of an object, not an array of objects (like [this example](https://json-schema.org/learn/miscellaneous-examples.html)).
Same thing will happen for `Device`.

When I wrote the first schemas, I had a key for each Device / ElectrodeGroup, instead of an array for all of them. But it seems like we’ve moved from that to have arrays in a single key instead, I guess it makes sense. 
I’ll make a PR that changes this for `BaseRecordingExtractorInterface`

Reply:
https://github.com/catalystneuro/nwb-conversion-tools/pull/168